runtime/codert_vm/cnathelp.cpp

```diff
@@ -2448,50 +2448,59 @@ old_slow_jitNewInstanceImplAccessCheck(J9VMThread *currentThread)
-		if (!J9ROMCLASS_IS_PUBLIC(thisROMClass)) {
-			/* Default visbility means the method can be sent from anywhere in the method's package */
-			if (thisClass->packageID != callerClass->packageID) {
-illegalAccess:
-				buildJITResolveFrame(currentThread, J9_SSF_JIT_RESOLVE_RUNTIME_HELPER, parmCount);
-				J9UTF8 *classNameUTF = J9ROMCLASS_CLASSNAME(thisROMClass);
-				J9UTF8 *nameUTF = J9ROMMETHOD_NAME(romMethod);
-				J9UTF8 *sigUTF = J9ROMMETHOD_SIGNATURE(romMethod);
-				j9object_t detailMessage = currentThread->javaVM->internalVMFunctions->catUtfToString4(
-						currentThread,
-						J9UTF8_DATA(classNameUTF),
-						J9UTF8_LENGTH(classNameUTF),
-						(U_8*)".",
-						1,
-						J9UTF8_DATA(nameUTF),
-						J9UTF8_LENGTH(nameUTF),
-						J9UTF8_DATA(sigUTF),
-						J9UTF8_LENGTH(sigUTF));
-				addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSEXCEPTION, detailMessage);
-				goto done;
-			}
+ 	J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(defaultConstructor);
+	void *oldPC = buildJITResolveFrame(currentThread, J9_SSF_JIT_RESOLVE_RUNTIME_HELPER, parmCount);
+	IDATA checkResult = currentThread->javaVM->internalVMFunctions->checkVisibility(currentThread, callerClass, thisClass, thisClass->romClass->modifiers, J9_LOOK_REFLECT_CALL);
+	if (checkResult < J9_VISIBILITY_ALLOWED) {
+		if (VM_VMHelpers::immediateAsyncPending(currentThread)) {
+			goto done;
 		}
-		/* See if the method is visible */
-		if (!J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccPublic)) {
-			if (J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccPrivate)) {
-				/* Private methods are visible only within their declaring class */
+		if (VM_VMHelpers::exceptionPending(currentThread)) {
+			goto done;
+		}
+ illegalAccess:
+		J9UTF8 *classNameUTF = J9ROMCLASS_CLASSNAME(thisClass->romClass);
+		J9UTF8 *nameUTF = J9ROMMETHOD_NAME(romMethod);
+		J9UTF8 *sigUTF = J9ROMMETHOD_SIGNATURE(romMethod);
+		j9object_t detailMessage = currentThread->javaVM->internalVMFunctions->catUtfToString4(
+				currentThread,
+				J9UTF8_DATA(classNameUTF),
+				J9UTF8_LENGTH(classNameUTF),
+				(U_8*)".",
+				1,
+				J9UTF8_DATA(nameUTF),
+				J9UTF8_LENGTH(nameUTF),
+				J9UTF8_DATA(sigUTF),
+				J9UTF8_LENGTH(sigUTF));
+		setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSEXCEPTION, detailMessage);
+		goto done;
+	}
+
+	/* See if the method is visible - don't use checkVisibility here because newInstance
+	 * has different rules for protected method visibility.
+	 */
+	if (!J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccPublic)) {
+		if (J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccPrivate)) {
+			/* Private methods are visible only within their declaring class
+			 * unless the two classes are in the same nest (JDK11 and beyond).
+			 */
+#if defined(J9VM_OPT_VALHALLA_NESTMATES)
+			if (thisClass->nestHost != callerClass->nestHost)
```

Comment:
This is insufficient for the nest mates case as the `nestHost` may not have been loaded and validated yet.  It likely needs to call `loadAndVerifyNestHost()` if the nestHosts are null.

Reply:
At the start of the function, checkVisibility is called with thisClass and callerClass, which should ensure the loading, shouldn't it?

Reply:
Classes are either `public` or `package` access at the VM level.  Never `private` which is the only case that can cause nest validation to occur.

Reply:
Should be fixed now.